scalalib/src/scalafmt/ScalafmtModule.scala

```diff
@@ -18,13 +18,23 @@ trait ScalafmtModule extends JavaModule {
 
   def scalafmtVersion: T[String] = "1.5.1"
 
+  def scalafmtGroupId: T[String] = T {
+    val v = scalafmtVersion()
+    if (v.startsWith("0.") || v.startsWith("1.") || v == " 2.0.0-RC1")
+      "com.geirsson"
+    else
+      "org.scalameta"
+  }
+
+  def scalafmtScalaVersion: T[String] = "2.12.8"
```

Comment:
Hardcoded versions have a high risk to blow up eventually. Wouldn't it be better to leave it abstract, and let the user decide which version to use? In most cases, the user module also implements `ScalaModule`, so a good default could be `ScalaModule.scalaVersion`. We could add a hint in the scaladoc.

Reply:
Using `ScalaModule.scalaVersion` by default also seems problematic to me, because Scalafmt` is currently only published for Scala 2.11 and 2.12 but can be used to format code built (or cross-built) with Scala 2.10 or a 2.13 milestone/RC (or even with Dotty, I think). Also, probably a smaller problem, but there is no benefit to downloading Scalafmt and its dependencies twice (once for 2.11, once for 2.12) just because you have a project that cross-builds for Scala 2.11 and 2.12.

Just leaving this abstract avoids putting a brittle choice in mill itself, but I don't think it's very user friendly - I think the user will typically only care about choosing the Scalafmt version, not the Scala version used for running Scalafmt, so that's not very intuitive to have to configure that?

I think the better long term solution might be to use [scalafmt-dynamic](https://scalameta.org/scalafmt/docs/installation.html#standalone-library), as that now seems to be the recommended way to integrate with Scalafmt. I haven't looked into this in details so far, but from what I understand, if we do this, we will just use it like any library we depend on. For the Scalafmt version configuration, we could just document that people should set the `version` key in their `.scalafmt.conf` file (and scalafmt-dynamic will take care of resolving the correct Scalafmt version). Of course that implies adding another dependency to mill, and there may be other tradeoffs I haven't noticed.

Reply:
@olafurpg what do you think we should do here?

Reply:
scalafmt-dynamic is the recommended library API and it handles stuff like
- respecting "version" setting in .scalafmt.conf
- picking the correct organization name for different version (we moved from com.geirsson to org.scalameta in v2)
- simple and documented API

See docs https://scalameta.org/scalafmt/docs/installation.html#standalone-library

Most importantly, editor integrations like IntelliJ and Metals already use scalafmt-dynamic so it would be great to do the same in Mill. This ensure the same scalafmt version is used in the editor and the build.

The dynamic module still has a few remaining issues, primarily around concurrency, but I'm optimistic those will get fixed sometime this summer.

Reply:
@lihaoyi @lefou I've started to work on using scalafmt-dynamic, and it actually seems pretty easy to do. I can submit a pull request for that soon if you want.